src/hashes.ts

```diff
@@ -1,44 +1,76 @@
-import { makeClass } from "./utils/make-class";
 import { HashPrefix } from "./hash-prefixes";
-import { coreTypes } from "./types";
-import { parseBytes } from "./utils/bytes-utils";
 import * as createHash from "create-hash";
+import { Hash256 } from "./types/hash-256";
+import { BytesList } from "./serdes/binary-serializer";
 
-const Sha512Half = makeClass(
-  {
-    Sha512Half() {
-      this.hash = createHash("sha512");
-    },
-    statics: {
-      put(bytes) {
-        return new this().put(bytes);
-      },
-    },
-    put(bytes) {
-      this.hash.update(parseBytes(bytes, Buffer));
-      return this;
-    },
-    finish256() {
-      const bytes = this.hash.digest();
-      return bytes.slice(0, 32);
-    },
-    finish() {
-      return new coreTypes.Hash256(this.finish256());
-    },
-  },
-  undefined
-);
-
-function sha512Half(...args) {
+/**
+ * Class for hashing with SHA512
+ * @extends BytesList So SerializedTypes can write bytes to a Sha512Half
```

Comment:
Should really implement a bytes sink interface rather than extend from a BytesList